### PR TITLE
Use configuration attributes for artifact filtering in dependency resolution

### PR DESCRIPTION
**Pull request by bigdaz**
This PR removes the separate 'format' property on `Configuration` and instead allows a consumer to define the required artifact types via an 'artifactType' attribute. This attribute is automatically produced for each `ResolvedArtifact` based on the `type` property.

By removing the separate 'format' property, this provides a consistent mechanism for a consuming `Configuration` to specify all aspects of the variant required. However, the current implementation is hard-wired to only consider the 'artifactType' attribute.

At this stage, the artifact attributes are not modelled in any way in the producing configuration: we simply map existing artifact properties to a fixed set of attributes on `ResolvedArtifact`. For external dependencies this is likely to be close to what we want, since we are limited in the set of metadata available. For local dependencies we'll probably want to eventually support custom artifact attributes on the producing configuration. However I'd like to see this requirement driven by real use cases.

I think this is a good step forward for gradle/performance#201, and it should not be too difficult to merge the relevant changes from #887 as well as other commits for this issue (e.g. https://github.com/gradle/gradle/commit/4033969872cde5adb94a841c66a01a8d31cdbe76). 

I think this PR is OK to merge into master as-is (it's not a spike), or we could keep working on it until gradle/performance#201 is complete. Over to you @jjohannes .

### CODE REVIEW

**Comment by bigdaz** [comment]
@jjohannes The `AttributeContainer` passed into `ArtifactTransformer` isn't always taken directly from the consumer configuration. [Sometimes it is](https://github.com/gradle/gradle/blob/9df8a33922e252249a7cb36ec57f65498135e0b2/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java#L948-L948) but [for a 'configuration view' it is not](https://github.com/gradle/gradle/blob/9df8a33922e252249a7cb36ec57f65498135e0b2/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java#L915-L917).

In the end, this is about the `VisitedArtifactSet` contract, which allows selection of artifacts based on a set of attributes. So I'm not sure how we can remove the `AttributeContainer` parameter from `ArtifactTransformer` methods.

File: subprojects/core/src/main/java/org/gradle/api/artifacts/ResolvableDependencies.java [hunk 1]
```diff
@@ -60,12 +61,12 @@
...
 
     /**
-     * Returns a view of this set containing files with the given format.
+     * Returns a view of this set containing files matching the requested attributes.
      *
      * @since 3.3
      */
     @Incubating
-    FileCollection getFiles(String format);
+    FileCollection getFiles(Map<?, ?> attributes);
```
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/api/artifacts/ResolvableDependencies.java]: Alright, if we want to have this functionality as well (matching against more attributes than the one in the configuration), we need the `attributes` parameter everywhere.

**Comment by jjohannes** [comment]
@bigdaz Right. I missed that part. Then this is good to merge from my side.

File: subprojects/core/src/main/java/org/gradle/api/artifacts/ResolvableDependencies.java (hunk 1 above)
Comment (bigdaz) [inline subprojects/core/src/main/java/org/gradle/api/artifacts/ResolvableDependencies.java]: Yep, we absolutely require this functionality. It's a way of getting different artifacts for a configuration  _using exactly the same dependency graph_. This is what we're referring to as a "configuration view".

Note that I'm not very happy with this API, but anything else was very hard to use from the consumer build script. I'd be tempted to extract an immutable `Attributes` out of `AttributeContainer`, and then provide a convenient way to construct an `Attributes` instance.

**Pull request merged**